modules/xfeatures2d/src/daisy.cpp

```diff
@@ -0,0 +1,2057 @@
+ */
+class DAISY_Impl : public DAISY
+{
+
+public:
+    /** Constructor
+     * @param radius radius of the descriptor at the initial scale
+     * @param q_radius amount of radial range divisions
+     * @param q_theta amount of angular range divisions
+     * @param q_hist amount of gradient orientations range divisions
+     * @param mode computation of descriptors
+     * @param norm normalization type
+     * @param H optional 3x3 homography matrix used to warp the grid of daisy but sampling keypoints remains unwarped on image
+     * @param interpolation switch to disable interpolation at minor costs of quality (default is true)
+     * @param use_orientation sample patterns using keypoints orientation, disabled by default.
+     */
+    explicit DAISY_Impl(float radius=15, int q_radius=3, int q_theta=8, int q_hist=8,
+        int mode = DAISY::ONLY_KEYS, int norm = DAISY::NRM_NONE, InputArray H = noArray(),
+        bool interpolation = true, bool use_orientation = false);
+
+    virtual ~DAISY_Impl();
+
+    /** returns the descriptor length in bytes */
+    virtual int descriptorSize() const {
+        // +1 is for center pixel
+        return ( (m_rad_q_no * m_th_q_no + 1) * m_hist_th_q_no );
+    };
+    /** returns the descriptor type */
+    virtual int descriptorType() const { return CV_32F; }
+    /** returns the default norm type */
+    virtual int defaultNorm() const { return NORM_L2; }
+
+    // main compute routine
+    virtual void compute( InputArray image, std::vector<KeyPoint>& keypoints, OutputArray descriptors );
+
+protected:
+
+    /*
+     * DAISY parameters
+     */
+
+    // operation mode
+    int m_mode;
+
+    // maximum radius of the descriptor region.
+    float m_rad;
+
+    // the number of quantizations of the radius.
+    int m_rad_q_no;
+
+    // the number of quantizations of the angle.
+    int m_th_q_no;
+
+    // the number of quantizations of the gradient orientations.
+    int m_hist_th_q_no;
+
+    // holds the type of the normalization to apply; equals to NRM_PARTIAL by
+    // default. change the value using set_normalization() function.
+    int m_nrm_type;
+
+    // holds optional H matrix
+    Mat m_h_matrix;
+
+    // input image.
+    float* m_image;
+
+    // image height
+    int m_h;
+
+    // image width
+    int m_w;
+
+    // stores the descriptors : its size is [ m_w * m_h * m_descriptor_size ].
+    float* m_dense_descriptors;
+
+    // stores the layered gradients in successively smoothed form: layer[n] =
+    // m_gradient_layers * gaussian( sigma_n ); n>= 1; layer[0] is the layered_gradient
+    float* m_smoothed_gradient_layers;
+
+    // if set to true, descriptors are scale invariant
+    bool m_scale_invariant;
+
+    // if set to true, descriptors are rotation invariant
+    bool m_rotation_invariant;
+
+    // number of bins in the histograms while computing orientation
+    int m_orientation_resolution;
+
+    // hold the scales of the pixels
+    float* m_scale_map;
```

Comment:
m_smoothed_gradient_layers is now Mat. It's the most important matrix of floats in daisy. Unfortunate is that is N dimensions (i think N > 5) so in many place i was able to use it as .ptr() reference without undersanding its internal store system. Speed was not affected by conversion at all. [DONE]